### PR TITLE
Defer handshake timeout in case the event loop is completely blocked

### LINKED ISSUE #21: Moderate to high concurrency load causes "Failed to launch process" errors on Windows
Using Artax (`composer require amphp/artax`) and running the following code on Windows:

```php
        Loop::run(function () {
            foreach (range(1, 1000) as $id) {
                $this->client->request('http://example.com')->onResolve(function ($error, $response) {
                    if ($error) {
                        echo "$error\n";

                        return;
                    }
                });
            }
        });
```

Generates the following error:

>Amp\Process\ProcessException: Process did not connect to server before timeout elapsed in vendor\amphp\process\lib\Internal\Windows\SocketConnector.php:344

The key is to have a high concurrency limit (i.e. 1000 in this case). Lower limits can trigger trigger it, too, but they need to be at least in the region of ~500 on my Windows 7 system.

The error is strange because only a couple of seconds are passing, at most, before the errors start streaming in. The same code works (can successfully download a response) with lower concurrency limits (e.g. around 200 or so).

----

Edit: After further testing, I can trigger this with a limit as low as 50. It seems to depend on how long the destination server takes to respond. Loading larger pages is more likely to cause it to fail. Again, only a second or so of real user time is passing but even trying to load 50 heavy URLs at once will cause it to fail the majority of the time and there seems to be no way to override the timeout.

### PR DESCRIPTION
**Pull request by kelunik**
Fixes #21.

@Bilge please try with this branch.

### CODE REVIEW

File: lib/Internal/Windows/SocketConnector.php [hunk 1]
```diff
@@ -353,7 +365,19 @@ public function onProcessConnectTimeout($watcher, Handle $handle) {
     }
 
     public function registerPendingProcess(Handle $handle) {
-        $handle->connectTimeoutWatcher = Loop::delay(self::CONNECT_TIMEOUT, [$this, 'onProcessConnectTimeout'], $handle);
+        // See onProcessConnectTimeout(), use double defer(), because defer() is always executed before timers.
+        $loopTicked = &$handle->loopTicked;
+        Loop::defer(static function () use (&$loopTicked) {
```
Comment (Bilge) [inline lib/Internal/Windows/SocketConnector.php]: What effect does `static` have here?

Comment (kelunik) [reply lib/Internal/Windows/SocketConnector.php]: It prevents `$this` being captured and thus the event loop having a reference to the object.

**Comment by Bilge** [comment]
Unfortunately this appears to have had no effect. Possible concurrency before failing due to timeout is exactly the same as before this patch.

**Comment by kelunik** [comment]
@Bilge Does the new patch change anything?

**Comment by Bilge** [comment]
Yes! :ok_hand: I can now achieve concurrency up to about ~1500 without changing the timeout. Anything more than that and I just run out of memory, but timeouts no longer seem to occur.

**Comment by Bilge** [comment]
Aside, I'm not sure force pushing is really such a good idea. Obviously your older solution was not the answer but now there's no record it ever existed. It's usually a good idea to keep old ideas around if only for posterity.

**Comment by kelunik** [comment]
@Bilge Nobody needs a non-working fix for an issue. ;-)

It's basically the same idea, but applied a bit differently.

**Comment by Bilge** [comment]
One day you might want to refer back to something you tried that didn't work. Just a thought; I don't mean to tell you how to run your projects.

**Comment by kelunik** [comment]
@Bilge There are cases where keeping the history is beneficial, but not in this case. The new implementation is both more elegant and actually correct. ;-)

**Comment by Bilge** [comment]
@kelunik Is it not possible to write a test case for this? It would be unfortunate if this were to creep back in as a regression later.

**Comment by kelunik** [comment]
@Bilge Definitely! Do you want to provide a PR? Basically it just needs to start a process and then directly call `sleep(2);` before yielding `$process->join();`.

**Comment by Bilge** [comment]
That doesn't sound as difficult as I was imagining, but I'm still not sure it is within my ability.

**Comment by kelunik** [comment]
@Bilge Just push a PR with what you have and we can improve it together. :-)

**Pull request merged**